python-checks/src/main/java/org/sonar/python/checks/PandasDataFrameToNumpyCheck.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.checks;
+
+import java.util.Optional;
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.QualifiedExpression;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.python.cfg.fixpoint.ReachingDefinitionsAnalysis;
+
+@Rule(key = "S6741")
+public class PandasDataFrameToNumpyCheck extends PythonSubscriptionCheck {
+
+  private static final String MESSAGE = "Do not use \"DataFrame.values\".";
+  private ReachingDefinitionsAnalysis reachingDefinitionsAnalysis;
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.FILE_INPUT, ctx -> reachingDefinitionsAnalysis = new ReachingDefinitionsAnalysis(ctx.pythonFile()));
+    context.registerSyntaxNodeConsumer(Tree.Kind.QUALIFIED_EXPR, this::checkForDataFrameValues);
+  }
+
+  private void checkForDataFrameValues(SubscriptionContext ctx) {
+    QualifiedExpression expr = (QualifiedExpression) ctx.syntaxNode();
+    if ((Optional.of(expr)
+      .filter(ex -> "values".equals(ex.name().name())).isEmpty())) {
+      return;
+    }
+
+    if (expr.qualifier().is(Tree.Kind.NAME)) {
+      this.reachingDefinitionsAnalysis.valuesAtLocation((Name) expr.qualifier())
+        .stream()
+        .filter(exp -> exp.is(Tree.Kind.CALL_EXPR))
+        .map(CallExpression.class::cast)
+        .filter(ce -> ce.callee().is(Tree.Kind.NAME, Tree.Kind.QUALIFIED_EXPR))
+        .map(PandasDataFrameToNumpyCheck::getFullyQualifiedName)
+        .filter(Optional::isPresent)
+        .map(Optional::get)
+        .filter("pandas.DataFrame"::equals).findAny()
```

Comment:
I have the feeling this is a bit too narrow, as there are many calls that may produce a `pandas.DataFrame` object besides the constructor (e.g `read_csv`) and that in the real world, most people won't actually use the constructor itself.

What we would really want to do is to rely on type inference to catch qualified expressions whose qualifier has type `pandas.DataFrame`, but because we don't have Pandas stubs yet, we won't understand which method/function is susceptible from returning this type. Maybe we can discuss considering adding such stubs as a separate ticket for this sprint, though.

Reply:
Yes, this is true and I agree. I decided to stay in this narrow scope exactly because I wasn't sure about the return type of functions such as `read_csv`. I agree that we should improve this later, and it would be great to have a meeting to discuss this.

Reply:
That sounds good. Could you create a ticket to expand this rule to common Pandas methods? We can keep the limited scope as part of this PR, but I think it would be nice to have another that expands on this during the sprint.

Reply:
Yes, I just created the ticket, which can be found [here](https://sonarsource.atlassian.net/browse/SONARPY-1488).